### DISCUSSION TITLE
1.4.0 CDN Changes and Semantic Versioning Guidance

### DISCUSSION THREAD

**Discussion by jameslittle230** [jameslittle230/stork#238] [Announcements]
Hello!

Previously, [I stopped publishing command line binaries](https://github.com/jameslittle230/stork/discussions/180) to the Stork CDN (`files.stork-search.net`) because it was causing too many backwards compatibility issues. This announcement is a follow-up to that post.

Starting in 1.4.0, I will no longer be publishing any new files to the root of the CDN, so as to not unintentionally break integrations. This means that, to upgrade from Stork 1.3.0 to Stork 1.4.0, you will need to update the script tag URLs and the CSS URLs as well as the binary download:

```html
<link rel="stylesheet" href="https://files.stork-search.net/releases/v1.4.0/edible.css" />
<script src="https://files.stork-search.net/releases/v1.4.0/stork.js"></script>
```

Each published version of `stork.js` will point to a pinned version of `stork.wasm`.

When 1.4.0 is published, the files on the CDN will be organized in the following way:

```
files.stork-search.net
│
│ // Files at the root will no longer be updated. They will point to the build artifacts
│ // from the 1.3.0 release and will eventually be deleted.
├─ stork.js
├─ stork.wasm
├─ basic.css
├─ dark.css
│ 
└─ releases
   ├─ v1.4.0
   │  ├─ stork-ubuntu-16-04
   │  ├─ stork-ubuntu-20-04
   │  ├─ stork-macos-10-15
   │  ├─ stork.js
   │  ├─ stork.wasm
   │  ├─ basic.css
   │  └─ dark.css
   │
   ├─ v1.3.0
   │  ├─ stork-ubuntu-16-04
   │  ├─ stork.js
   │  └─ ...
   │
   └─ latest // Files in this directory will still be updated when there is a Stork release.
      ├─ stork-ubuntu-16-04
      ├─ stork.js
      └─ ...
```

---

I also want to document which aspects of the project adhere to the semantic versioning rules, and which do not. I will maintain semantic versioning adherence in the following areas:

- Any working configuration file will still work with any later version of the Stork indexing binary, up to the next major version
- Any command line invocation will still give the same output for any later binaries, up to the next major version
- Any working calls to the Stork JS API (`.register()`, .initialize()`, .search()` and others) will still work with any later versions of `stork.js`, up to the next major version
- Any index file can be loaded with any later versions of `stork.js`/`stork.wasm`, up to the next major version

Similarly, I _cannot_ guarantee semantic versioning adherence in the following areas (although if Stork is causing major issues with these, please reach out):

- The API of the Rust `stork-search` library published to crates.io
- HTML and CSS side effects of Stork themes
- Development processes

---

Please let me know if you have any suggestions or questions!

**Comment by ArsenArsen** [comment]
hey, version 1.4.0 appears to not be available on crates.io, is this intentional?

**Reply by jameslittle230** [reply]
Thanks for the nudge here - I knew it wasn't up yet, but 1.4.0 ended up being trickier to publish on Crates.io and I hadn't gotten around to fixing it. I'll take another stab at this and try pushing 1.4.0 again as soon as I can.

## The nitty gritty

In 1.4.0, I split Stork into a bunch of independent libraries for ease of development. I don't (yet) want to publish each crate independently on Crates.io, so I have to figure out if I can specify that when I publish the Stork binary to Crates.io, I want to vendor those dependencies. At first glance, it wasn't possible, but I may have not been looking hard enough.

**Reply by ArsenArsen** [reply]
Ah, I missed that. Apologies!

**Reply by jameslittle230** [reply]
@ArsenArsen 1.4.1 has been published to Crates.io :)
